### PR TITLE
refs #3024 fixed a bug where call references did not set the executio…

### PR DESCRIPTION
**Pull request by davidnich**
…n context with builtin functions and therefore calls to builtin functions in modules with per-program private data would fail

test is in the jni module; requires a binary module with private data to test

### CODE REVIEW

File: lib/QoreLib.cpp [hunk 1]
```diff
@@ -2663,4 +2663,8 @@ QoreHashNode* get_source_location(const QoreProgramLocation* loc) {
 
 const char* type_get_name(const QoreTypeInfo* t) {
     return QoreTypeInfo::getName(t);
+}
+
+QoreProgram* qore_get_call_program_context() {
+    return get_program_call_context();
 }
```
Comment (sejvlond) [inline lib/QoreLib.cpp]: is it necessary to create new function, if there is one doing the same?

File: include/qore/intern/qore_thread_intern.h [hunk 2]
```diff
@@ -363,6 +367,22 @@ DLLLOCAL QoreListNode* getCallStackList();
 #define popCall(x)
 #endif
 
+class ProgramCallContextHelper {
+public:
+    DLLLOCAL ProgramCallContextHelper(QoreProgram* new_pgm)
+        : pgm(new_pgm ? get_set_program_call_context(new_pgm) : reinterpret_cast<QoreProgram*>(-1)) {
```
Comment (sejvlond) [inline include/qore/intern/qore_thread_intern.h:372]: are u sure about reinterpret cast -1 into a pointer?

if I understand it correctly, you want to set new pgm and then return the old one, unless pgm is null, then you want to keep the original one. So cannot we simply do
```
pgm(new_pgm ? get_set_program_call_context(new_pgm) : get_program_call_context()) 
```
and then in destructor always set back the one that was there?

File: lib/QoreLib.cpp (hunk 1 above)
Comment (davidnich) [reply lib/QoreLib.cpp]: fixed

File: include/qore/intern/qore_thread_intern.h (hunk 2 above)
Comment (davidnich) [reply include/qore/intern/qore_thread_intern.h:372]: as discussed yesterday

**Review by sejvlond — APPROVED** [review]

**Pull request merged**